### PR TITLE
Cli

### PR DESCRIPTION
**Pull request by kecajjo**

### CODE REVIEW

**Comment by kecajjo** [comment]
Why CI didnt start this time?

**Comment by ProrokWielki** [comment]
> Why CI didnt start this time?

As far as i can see it was run for all the commits in this PR

**Comment by ProrokWielki** [comment]
Split this PR into 2 - one introducing conan - without ncurses being used, and the second one adding ncurses drawing engine

File: CMakeLists.txt [hunk 1]
```diff
@@ -8,10 +8,12 @@ set(TIDY
 set(CMAKE_CXX_COMPILER g++)
 set(CMAKE_CXX_STANDARD 17)
 
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -pedantic -Wextra")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -pedantic -Wextra -g")
```
Comment (ProrokWielki) [inline CMakeLists.txt:11]: if you are adding debug i would do it right, have one config for debug - no debug symbols and no optimization and release config with o2 or whatever you see fit

File: gui/color/inc/color.hpp [hunk 2]
```diff
@@ -9,4 +9,4 @@ struct Color {
     Color(const uint8_t& red, const uint8_t& green, const uint8_t& blue, const uint8_t& alpha)
         : r(red), g(green), b(blue), a(alpha) {}
     static constexpr unsigned max_color_val = std::numeric_limits<unsigned char>::max();
-};
\ No newline at end of file
+};
```
Comment (ProrokWielki) [inline gui/color/inc/color.hpp:12]: could you force your editor or add pre-commit hook or whetever so you always or never have a new line at the and of the file :P

File: main/main.cpp [hunk 3]
```diff
@@ -20,4 +33,15 @@ int main() {
...
+    std::shared_ptr<Board> b = std::make_shared<Board>();
+    Drawable::Position p{0, 0};
+    for (auto &s : v) {
+        b->addBlock(std::move(s), p);
+        p = p + Drawable::Position{xx, yy};
+    }
+    std::shared_ptr<DrawingEngine> de = std::make_shared<CliColor>(height, width);
+    de->drawDrawable(b, {0, 0});
+    de->display();
+    sleep(sleepsec);
```
Comment (ProrokWielki) [inline main/main.cpp]: sleep really? std::this_thread::sleep_for

File: gui/drawing_engine/cli_color/src/cli_color.cpp [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    return color;
+}
+
+}  // namespace
+
+CliColor::CliColor(const unsigned int& height, const unsigned int& width)
+    : DrawingEngine(height, width) {
+    initscr();
+    if (!has_colors()) {
+        throw std::runtime_error("Terminal doesnt support colors");
```
Comment (ProrokWielki) [inline gui/drawing_engine/cli_color/src/cli_color.cpp:39]: i'd go in some retarded mode and not fail completely :P

File: gui/drawing_engine/cli_color/src/cli_color.cpp [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    }
+    start_color();
+    setupColors();
+}
+
+CliColor::~CliColor() { endwin(); }
+
+void CliColor::drawFrameBuffer() {
+    unsigned int height = getHeight();
```
Comment (ProrokWielki) [inline gui/drawing_engine/cli_color/src/cli_color.cpp]: const

**Comment by ProrokWielki** [comment]
As for the next steps, you could also add the IWYU to your CI and the `ftxui` for next Drawing engine implementation and if you want to try real GUI you can try using opengl. It would be nice if your program could take parameters to choose which drawing engine to use

**Comment by kecajjo** [comment]
> As for the next steps, you could also add the IWYU to your CI and the `ftxui` for next Drawing engine implementation and if you want to try real GUI you can try using opengl. It would be nice if your program could take parameters to choose which drawing engine to use

I'll do drawing engine factory which will return DE that I'll be using, I'll make each concrete DE private include of the factory lib and as a last part (probably last thing done in the project) Ill dynamically load DE libs and load only the one I'll be using with other option used in case of failed load. For now I'll stick with 2 DE (I'll add `ftxui`) and just do factory for them laster

**Comment by kecajjo** [comment]
I'll rebase it after the other PR is merged, than CI will pass

**Review by ProrokWielki — CHANGES_REQUESTED** [review]

File: gui/drawing_engine/cli_color/src/cli_color.cpp (hunk 4 above)
Comment (kecajjo) [reply gui/drawing_engine/cli_color/src/cli_color.cpp:39]: Once there is a non color version of CLI I'll handle the exception in a DE factory, but i need other DE for this

File: CMakeLists.txt (hunk 1 above)
Comment (kecajjo) [reply CMakeLists.txt:11]: This PR is big, I'll change it in the next PR

File: gui/color/inc/color.hpp (hunk 2 above)
Comment (kecajjo) [reply gui/color/inc/color.hpp:12]: I have to find a way to do this :P

File: main/main.cpp (hunk 3 above)
Comment (kecajjo) [reply main/main.cpp]: fixed

File: gui/drawing_engine/cli_color/src/cli_color.cpp (hunk 5 above)
Comment (kecajjo) [reply gui/drawing_engine/cli_color/src/cli_color.cpp]: fixed

Comment (ProrokWielki) [reply gui/drawing_engine/cli_color/src/cli_color.cpp]: i dont see the change ... i meant `height` and `width` should be const

File: main/main.cpp [hunk 6]
```diff
@@ -1,13 +1,28 @@
...
 #include <iostream>
 #include <memory>
+#include <thread>
+#include <chrono>
 #include <vector>
 
+#include "board.hpp"
+#include "cli_color.hpp"
 #include "diagonal.hpp"
+#include "drawable.hpp"
```
Comment (ProrokWielki) [inline main/main.cpp]: thats a lot of includes :P You should create shape factory and only include it :P

**Review by ProrokWielki — CHANGES_REQUESTED** [review]

File: main/main.cpp (hunk 6 above)
Comment (kecajjo) [reply main/main.cpp]: It was done in the other PR but i didn't rebase yet

File: gui/drawing_engine/cli_color/src/cli_color.cpp (hunk 5 above)
Comment (kecajjo) [reply gui/drawing_engine/cli_color/src/cli_color.cpp]: Oh, I've made the whole method const, fixing height and width this time

File: gui/color/inc/color.hpp (hunk 2 above)
Comment (kecajjo) [reply gui/color/inc/color.hpp:12]: I'll always have a new line from now on

**Review by ProrokWielki — APPROVED** [review]

**Pull request merged**